[REFACTOR!] Version 2.0.0 proposal

leo-ls opened the pull request:
Hi @preetamkajalrout,

Your middleware is perhaps the most useful I've found. It allows for a full development experience out of WebIDE. Thanks for publishing it.

To contribute to your project, I'm opening this PR with a big refactor, so the project can be even more useful.

The highlights of the refactor are:
- Instead of a custom JSON file, I've implemented the use of WebIDE destination files;
- I've also added handling for CSRF tokens and cookies, so requests are now fully authenticated against OData sources;
- I've changed some of the resource serving logic, so local resources are preferred.

Kindly take some time to review the code and let me know what you think.

Best regards,
Leonardo

preetamkajalrout commented:
Hi @leo-ls ,

First of all, changes are indeed breaking and happy to see ES6 style coding in the code-base (really welcome it). I gave it a look and few things that jump out at this point:

1. Support for local WebIDE like destination config support (property files):
- This should be added as an alternative feature rather than only feature.
- Support for existing json-store should not be removed. Rational behind it would be to support custom config in future like headers, request/response modification w/o touching neo-app.json

_I agree to the objective of supporting properties file, this adds a lot of value to the middleware and would help users move from Personal WebIDE to ui5-tooling instantly. However, both should exist as alternatives_ 

2. Removal of **@ui5/cli** from dependency. Can you help me with the reason?

I haven't had time to look into the code in depth. Will go through that, and add comments anywhere needed.

On a seperate note, Changes made are really appreciated, Thanks for your effort, I can see this change will add a lot of value to users. I am inclined to merge this to main repo.

leo-ls commented:
> 1. Support for local WebIDE like destination config support (property files):
> 
> * This should be added as an alternative feature rather than only feature.
> * Support for existing json-store should not be removed. Rational behind it would be to support custom config in future like headers, request/response modification w/o touching neo-app.json
> 
> _I agree to the objective of supporting properties file, this adds a lot of value to the middleware and would help users move from Personal WebIDE to ui5-tooling instantly. However, both should exist as alternatives_

Completely agree, sorry for missing the point. I refactored that part in the last commits to allow again for the json file.

Also, I changed its structure a little, please review it.

> 2. Removal of **@ui5/cli** from dependency. Can you help me with the reason?

As this is a server middleware, I understand that it can only run in the context of UI5 Tooling, so `@ui5/cli` (or maybe only `@ui5/server`) is already an upstream dependency and available if needed. Therefore, I don't think it's necessary to declare it explicitly as a dependency of this package. What are your thoughts on this?

Thanks for the reply. Looking forward to seeing this on master!

preetamkajalrout commented on src/main.js:
```diff
@@ -1,175 +1,95 @@
-      addHeader(res, HEADER_ALLOW_CONTROL_ALLOW_ORIGIN, "*");
-      addHeader(res, HEADER_ALLOW_CONTROL_ALLOW_CREDENTIALS, "true");
-      addHeader(res, HEADER_ALLOW_CONTROL_ALLOW_METHODS, "GET,PUT,POST,DELETE");
-      addHeader(res, HEADER_ALLOW_CONTROL_ALLOW_HEADERS, "X-Requested-With, Accept, Origin, Referer, User-Agent, Content-Type, Authorization, X-Mindflash-SessionID");
-      // Intercept OPTIONS method
-      if (req.method === "OPTIONS") {
-        // addHeader(res, HEADER_ALLOW_CONTROL_ALLOW_ORIGIN, "*");
-        res.status(200);
-        next();
-        return;
-      }
+      } else if (serveWithProxy) {
```
There is a use-case where if user doesn't provide any configuration for SAPUI5 resources in ui5.yml . None of the if/elseif condition will proxy for **route** of _type_ **service**.

preetamkajalrout commented:
Hi @leo-ls ,

Thanks for the changes. It absolutely, handles the json file as expected. 
I have provided some review comments, Can you resolve them. Would merge the changes to master after that.

I see there is gap in use-case in the solution. If user doesn't provide any config for SAPUI5 resources in ui5.yml . Then, it is not being handled currently, since the **type** of _route_ in neo-app.json will be **service**. Suggested change would be to consider a destination entry in json store if ui5.yml & environment variable are not set:

```jsonc
{
 "Name": "sapui5",
 "Description": "SAPUI5 Resources",
 "customPath": "C:\\Dev\\sapui5\sdk", // OR: https://hostname:port
 "preferLocal": true, // OR: false
 "strictSSL": true // OR: false
}
```
Following are the reason behind the change:

1.   **customPath** can be different for multiple developers working on same UI5 app. In turn, changing ui5.yml after each clone result in overhead
2. Developer can use the same config (SAPUI5 resources) for multiple apps they are working on.

> As this is a server middleware, I understand that it can only run in the context of UI5 Tooling, so @ui5/cli (or maybe only @ui5/server) is already an upstream dependency and available if needed. Therefore, I don't think it's necessary to declare it explicitly as a dependency of this package. What are your thoughts on this?

I agree on this. This is not required for the middleware. Thanks for the catch!

Please let me know your views on these changes.

preetamkajalrout commented:
Hi @leo-ls ,

I have taken some code of yours and made some changes on top of it to support some use-cases & also support Node 10+. However, I haven't merged the pull request as some of the use-cases were not working. For e.g. static files (project resources) were failing.

I will add the feature of properties as you have already coded in few days. 
I have also added your name & reference to Contributors list in [README](https://github.com/preetamkajalrout/ui5-middleware-destination#contributors). Thanks for your contribution! Really appreciate it 👍
